nomad/leader.go

```diff
@@ -1327,3 +1329,25 @@ func (s *Server) getOrCreateSchedulerConfig() *structs.SchedulerConfiguration {
 
 	return config
 }
+
+func (s *Server) getOrCreateClusterID() string {
+	fsmState := s.fsm.State()
+	existingMeta, err := fsmState.ClusterMetadata()
+	if err != nil {
+		s.logger.Named("core").Error("failed to get cluster ID", "error", err)
+		return ""
+	}
+
+	if existingMeta == nil || existingMeta.ClusterID == "" {
+		newMeta := structs.ClusterMetadata{ClusterID: uuid.Generate()}
+		if _, _, err = s.raftApply(structs.ClusterMetadataRequestType, newMeta); err != nil {
```

Comment:
Good catch! We definitely need to use minimum version to avoid the case:

1. Server 1 (of 3) upgraded, elected leader, and sets ID
2. Servers 2 & 3 ignore ID
3. Server2 2 & 3 are upgraded, has no cluster ID
4. Server 1 crashes, Server 2 elected leader, sets *new* ID.

We now have a cluster with 2 cluster IDs.

Reply:
Does `IgnoreUnknownTypeFlag` address this issue?  Servers 2&3 will ignore the ID in step 2, and when upgraded they will replay log and all should be fine.  But I don't know how it will cope with snapshots (e.g. if Server 2 running old version creates a snapshot without interpreting cluster id log, will upgraded Server 2 effectively drop reviewing that log entry?)'.

With a minimum version check, `getOrCreateClusterID` will need to be called for each access outside the leader loop to handle the following case:

1. Server 1 (of 3) upgraded and elected leader, doesn't set an ID yet
2. Server 2 & 3 are upgraded sequentially such that Server 1 remains as leader
3. All servers are fully upgraded but `establishLeadership` isn't called again and no cluster ID set.

Reply:
> Does IgnoreUnknownTypeFlag address this issue? Servers 2&3 will ignore the ID in step 2, and when upgraded they will replay log and all should be fine. But I don't know how it will cope with snapshots (e.g. if Server 2 running old version creates a snapshot without interpreting cluster id log, will upgraded Server 2 effectively drop reviewing that log entry?)'.

I don't know of how IgnoreUnknownTypeFlag could cope with snapshotting. I think that's a big part of the reason it's rarely used because there's many cases in which in could cause state to diverge forever.